File: Sales_Billing/src/dplyr/report_KE30_and_GL.R
library(dplyr)
library(readr)
library(stringr)
library(janitor)
library(here)
library(glue)


# Path
path <- here("Sales_Billing")


# Functions
process_ke30 <- function(df) {
  df <- df |>
    mutate(
      across(c("quantity", "revenues", "stock_val"), as.double)
    ) |>
    select(
      c(
        "period",
        "profit_center",
        "product",
        "quantity",
        "stock_val",
        "revenues"
      )
    ) |>
    mutate(
      period = str_sub(.data$period, 7, 8),
      source = "KE30", .before = "period"
    )
  return(df)
}


process_gl <- function(df) {
  df <- df |>
    mutate(
      across(
        c("quantity", "amount_in_doc_curr", "amount_in_local_cur"),
        as.double
      )
    ) |>
    select(c(
      "period",
      "profit_ctr",
      "material",
      "quantity",
      "amount_in_local_cur"
    )) |>
    rename(
      "profit_center" = "profit_ctr",
      "product" = "material",
      "stock_val" = "amount_in_local_cur"
    ) |>
    mutate(source = "GL Account", .before = "period")

  return(df)
}


summary_data <- function(df) {
  df <- df |>
    group_by(
      across(c("source", "period", "profit_center", "product")),
    ) |>
    summarise(
      quantity = sum(.data$quantity),
      sales = sum(.data$revenues),
      costs = sum(.data$stock_val),
      .groups = "drop"
    )
  return(df)
}


main <- function() {
  # Variables
  year <- "2024"

  # Filenames
  input_1 <- here(path, "db", glue("KE30_{year}.csv"))
  input_2 <- here(path, "db", glue("GL_{year}.csv"))
  output_file <- here(path, "output", glue("KE30 and GL Account_{year}.csv"))

  # Read data
  df_ke30 <- read_csv(input_1, col_types = list(.default = col_character()))
  df_gl <- read_csv(input_2, col_types = list(.default = col_character()))

  # Process data
  df_ke30 <- df_ke30 |> process_ke30()
  df_gl <- df_gl |> process_gl()
  df <- bind_rows(df_ke30, df_gl)

  df <- df |>
    summary_data() |>
    arrange(desc(source))

  # Write data
  write_csv(df, output_file)
  print("A file is created")
}

main()

File: Sales_Billing/src/pandas/report_KE30_and_GL.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent.parent


# Functions
def process_ke30(df):
    # select columns
    selected_columns = [
        "period",
        "profit_center",
        "product",
        "quantity",
        "stock_val",
        "revenues",
    ]
    df = df[selected_columns]

    # text slicing & convert data type
    df["period"] = df["period"].str[-2:]  # extract the last two characters
    df["period"] = df["period"].astype(int)

    # add column
    df["source"] = "KE30"

    # reorder columns
    df = df[["source"] + [col for col in df.columns if col not in ["source"]]]

    return df


def process_gl(df):
    # select columns
    selected_columns = [
        "period",
        "profit_ctr",
        "material",
        "quantity",
        "amount_in_local_cur",
    ]
    df = df[selected_columns]

    # rename columns
    df = df.rename(
        columns={
            "profit_ctr": "profit_center",
            "material": "product",
            "amount_in_local_cur": "stock_val",
        }
    )

    # add column
    df["source"] = "GL"

    # reorder columns
    df = df[["source"] + [col for col in df.columns if col not in ["source"]]]

    return df


def summary_data(df):
    df = (
        df.groupby(
            [
                col
                for col in df.columns
                if col not in ["quantity", "revenues", "stock_val"]
            ]
        )
        .agg(
            qty=("quantity", "sum"),
            sales=("revenues", "sum"),
            costs=("stock_val", "sum"),
        )
        .reset_index()
    )
    return df


def main():
    # Variables
    year = "2024"

    # Filenames
    input_1 = path / "db" / f"KE30_{year}.csv"
    input_2 = path / "db" / f"GL_{year}.csv"
    output_file = path / "output" / f"KE30 and GL Account_{year}.csv"

    # Read data
    df_ke30 = pd.read_csv(
        input_1,
        dtype={
            # text
            "period": str,
            # number
            "revenues": int,
            "stock_val": int,
            "quantity": int,
        },
    )
    df_gl = pd.read_csv(
        input_2,
        dtype={
            # number
            "quantity": int,
            "amount_in_doc_curr": int,
            "amount_in_local_cur": int,
        },
    )

    # Process data
    df_ke30 = process_ke30(df_ke30)
    df_gl = process_gl(df_gl)

    df = pd.concat([df_ke30, df_gl])
    df["revenues"] = df["revenues"].fillna(0).astype(int)

    df = df.pipe(summary_data).sort_values(
        by=["source", "period", "profit_center", "product"],
        ascending=[False, True, True, True],
    )

    # Write data
    df.to_csv(output_file, index=False)
    print("A file is created")


if __name__ == "__main__":
    main()
